### ISSUE TITLE
Color Gamut Mapping in CIECAM02 preserving Hue Angle

### ISSUE THREAD

**Issue by briend** [briend/briend.github.io#4]
General Problem:  Take the color appearance attributes of one color and shift the hue while maintaining the other color appearance attributes as much as possible.  When necessary, reduce the J/Q (lightness/brightness) and/or the M/C/s (Colorfulness/Chroma/Saturation) until within the sRGB gamut as well as display limits of 0-1.0 (all RGB need to be 0-1.0).

Challenge:  We should minimize the reduction of both J/Q and M/C/s and find the solution that maximizes both of these attributes without exceeding the original values.  Probably using Python SciPy minimize function:
https://docs.scipy.org/doc/scipy/reference/optimize.minimize-lbfgsb.html

Basically, very similar to this code, but NOT using a uniform color space:
https://github.com/crowsonkb/color_schemer/blob/master/cam.py#L73-L84

Yellow:
colorspacious.cspace_convert([1,1,0], "sRGB1", "JMh")
array([ 95.67306142, 58.26474923, 106.14599451])
![image](https://user-images.githubusercontent.com/6015639/39823056-052a6db0-5361-11e8-9ece-5d3bdea54f9e.png)

Blue:
colorspacious.cspace_convert([0,0,1], "sRGB1", "JMh")
array([ 21.074306 , 65.43521286, 257.9160105 ])
![image](https://user-images.githubusercontent.com/6015639/39822856-803d863c-5360-11e8-80e0-90e94768bcb2.png)

Now I start off with yellow's Lightness and Colorfulness,  but want to shift just the Hue to Blue:
colorspacious.cspace_convert([ 95.67306142, 58.26474923, 257.9160105], "JMh", "sRGB1")
array([0.69303055, 0.96741851, 1.57349333])
Since RGB > 1.0 we get a dramatic hue shift toward cyan since we are naive clamping to 1.0:
![image](https://user-images.githubusercontent.com/6015639/39823056-052a6db0-5361-11e8-9ece-5d3bdea54f9e.png) ![image](https://user-images.githubusercontent.com/6015639/39823886-dd017912-5362-11e8-9245-ba3ca79341eb.png)

If I ramp down just colorfulness we can preserve hue and lightness, but it's basically achromatic which is not terribly useful:
colorspacious.cspace_convert([ 95.67306142, 5.98, 257.9160105], "JMh", "sRGB1")
array([0.96006347, 0.96330082, 0.99997855])
![image](https://user-images.githubusercontent.com/6015639/39823056-052a6db0-5361-11e8-9ece-5d3bdea54f9e.png) ![image](https://user-images.githubusercontent.com/6015639/39824018-2babd1de-5363-11e8-9665-0d95b148bb51.png)

Or, I can ramp lightness and preserve hue and colorfullness.  This color is dramatically darker, however:
colorspacious.cspace_convert([ 35.5, 58.26474923, 257.9160105], "JMh", "sRGB1")
array([0.13161491, 0.39273014, 0.99957602])
![image](https://user-images.githubusercontent.com/6015639/39823056-052a6db0-5361-11e8-9ece-5d3bdea54f9e.png) ![image](https://user-images.githubusercontent.com/6015639/39824126-6bc75fb8-5363-11e8-84fb-1f491c541032.png)

A compromise might be to ramp BOTH lightness and colorfulness down while preserving hue angle. 

Here's a first stab, it seems to work.  I had to penalize chroma loss 10% to avoid nearly achromatic result
![image](https://user-images.githubusercontent.com/6015639/39823056-052a6db0-5361-11e8-9ece-5d3bdea54f9e.png) ![image](https://user-images.githubusercontent.com/6015639/39885393-522b70ba-5441-11e8-9bb9-316a21eecc32.png)

![image](https://user-images.githubusercontent.com/6015639/39822856-803d863c-5360-11e8-80e0-90e94768bcb2.png)![image](https://user-images.githubusercontent.com/6015639/39885317-24e45a2c-5441-11e8-8f3d-ba557899ab27.png)

```python
import numpy as np
import colorspacious as cs
import scipy.optimize as spo

def constrain_to_gamut(JMh, penalize_chroma, tols):

  def apply_constraint(inputs):
      #rgb must >= 0 and <= 1
      rgb_ = cs.cspace_convert([inputs[0], inputs[1], JMh[2]], "JMh", "sRGB1")
      return min(min(rgb_), 1.0 - max(rgb_)) 

  def loss(JM_):
      #optionally penalize colorfulness loss to avoid achromatic results
      #although this reduces accuracy, technically
      loss = abs(JMh[0] - JM_[0]) + abs(JMh[1] - JM_[1]) * penalize_chroma
      return loss

  rgb = cs.cspace_convert(JMh, "JMh", "sRGB1")
  print("incoming rgb is", rgb)
  x = np.clip(rgb, 0, 1)
  if (rgb == x).all():
      return x

  #inital guess just cut them both in half
  guess = np.array([JMh[0] * .5 , JMh[1] * .5])
  #we shouldn't try lightness/colorfulness values that are impossible, and never increase colorfulness. 
  lowerlimit=cs.cspace_convert([0, 0, 0], "sRGB1", "JMh")
  upperlimit=cs.cspace_convert([1, 1, 1], "sRGB1", "JMh")
  bounds = ((lowerlimit[0], upperlimit[0]), (lowerlimit[1], JMh[1]))
  opt = {'disp':True}
  my_constraints = {'type': 'ineq', "fun": apply_constraint }

  x_opt = spo.minimize(loss,
                       guess,
                       method='SLSQP',
                       constraints=my_constraints,
                       bounds=bounds,
                       tol=tols,
                       options=opt
                      )

  answer = x_opt['x']
  print("final rgb is", cs.cspace_convert([answer[0], answer[1], JMh[2]], "JMh", "sRGB1"))  
  print(x_opt)                        
  return (answer[0], answer[1], JMh[2])

#blue hue w/ yellow's attributes
constrain_to_gamut([95.67306142, 58.26474923, 257.9160105], 1.1, None)
#yellow hue w/ blue's attributes
constrain_to_gamut([21.074306 , 65.43521286, 106.14599451], 1.1, None)

**Comment by briend** [comment]
First stab.. it doesn't work, but almost does, kinda

import numpy as np
import colorspacious as cs
import scipy.optimize as spo


def constrain_to_gamut(JMh):
    
    def apply_constraint(inputs):
        rgb = cs.cspace_convert([inputs[0], inputs[1], JMh[2]], "JMh", "sRGB1")
        print(inputs)
        if (rgb >= 0).all() and (rgb <= 1).all():
          print("true")
          return 1
        else:
          return -1

    my_constraints = {'type': 'ineq', "fun": apply_constraint }
    
    rgb = cs.cspace_convert(JMh, "JMh", "sRGB1")
    x = np.clip(rgb, 0, 1)
    if (rgb == x).all():
        return x

    def loss(JM_):
        loss = abs(JMh[0] - JM_[0]) + abs(JMh[1] - JM_[1])
        print(loss)
        return loss

    guess = np.array([JMh[0] * .5 , JMh[1] * .5])
    print(guess)
    bounds = ((0, JMh[0]), (0, JMh[1]))
    opt = {'disp':True, 'catol':100, 'rhobeg':1}

    x_opt = spo.minimize(loss,
                            guess,
                            method='COBYLA',
                            constraints=my_constraints,
                            )
    
    answer = x_opt['x']
    print(answer, JMh[2])
    print(cs.cspace_convert([answer[0], answer[1], JMh[2]], "JMh", "sRGB1"))  
    print(x_opt)                        
    return x_opt

constrain_to_gamut([95.67306142, 58.26474923, 257.9160105])

**Comment by briend** [comment]
This one works kinda.  I penalize the chroma loss otherwise it's pretty much achromatic 

import numpy as np
import colorspacious as cs
import scipy.optimize as spo

def constrain_to_gamut(JMh, penalize_chroma):

  def apply_constraint(inputs):
      #rgb must greater than 0 and less than 1
      rgb = cs.cspace_convert([inputs[0], inputs[1], JMh[2]], "JMh", "sRGB1")
      return min(min(rgb) + 1.0, 1.0 - max(rgb)) 

  my_constraints = {'type': 'ineq', "fun": apply_constraint }

  rgb = cs.cspace_convert(JMh, "JMh", "sRGB1")
  x = np.clip(rgb, 0, 1)
  if (rgb == x).all():
      return x

  def loss(JM_):
      #penalize colorfulness loss
      loss = abs(JMh[0] - JM_[0]) + abs(JMh[1] - JM_[1]) * penalize_chroma
      print(loss)
      return loss

  guess = np.array([JMh[0] * .5 , JMh[1] * .5])
  bounds = ((0, JMh[0]), (0, JMh[1]))
  opt = {'disp':False}

  x_opt = spo.minimize(loss,
                          guess,
                          method='SLSQP',
                          constraints=my_constraints,
                          bounds=bounds
                          )

  answer = x_opt['x']
  print(answer, JMh[2])
  print(cs.cspace_convert([answer[0], answer[1], JMh[2]], "JMh", "sRGB1"))  
  print(x_opt)                        
  return x_opt

constrain_to_gamut([95.67306142, 58.26474923, 257.9160105], 1.1)

**Comment by briend** [comment]
`import numpy as np
import colorspacious as cs
import scipy.optimize as spo

def constrain_to_gamut(JMh, penalize_chroma):

  def apply_constraint(inputs):
      #rgb must greater than 0 and less than 1
      rgb = cs.cspace_convert([inputs[0], inputs[1], JMh[2]], "JMh", "sRGB1")
      return min(min(rgb) + 1.0, 1.0 - max(rgb)) 

  my_constraints = {'type': 'ineq', "fun": apply_constraint }

  rgb = cs.cspace_convert(JMh, "JMh", "sRGB1")
  x = np.clip(rgb, 0, 1)
  if (rgb == x).all():
      return x

  def loss(JM_):
      #penalize colorfulness loss
      loss = abs(JMh[0] - JM_[0]) + abs(JMh[1] - JM_[1]) * penalize_chroma
      print(loss)
      return loss

  guess = np.array([JMh[0] * .5 , JMh[1] * .5])
  bounds = ((0, JMh[0]), (0, JMh[1]))
  opt = {'disp':False}

  x_opt = spo.minimize(loss,
                          guess,
                          method='SLSQP',
                          constraints=my_constraints,
                          bounds=bounds
                          )

  answer = x_opt['x']
  print(answer, JMh[2])
  print(cs.cspace_convert([answer[0], answer[1], JMh[2]], "JMh", "sRGB1"))  
  print(x_opt)                        
  return x_opt

constrain_to_gamut([95.67306142, 58.26474923, 257.9160105], 1.1)`
